frontend/src/app/worker/snaps.cljs

```diff
@@ -6,179 +6,31 @@
-              (update-in [frame-id :y] remove-coord-data frame-id shapes :y)))
-
-        add-data
-        (fn [snap-data [frame-id shapes]]
-          (-> snap-data
-              (update-in [frame-id :x] add-coord-data frame-id shapes :x)
-              (update-in [frame-id :y] add-coord-data frame-id shapes :y)))
-
-        delete-frames
-        (fn [snap-data frame-id]
-          (dissoc snap-data frame-id))
-
-        add-frames
-        (fn [snap-data frame-id]
-          (assoc snap-data frame-id {:x (rt/make-tree)
-                                     :y (rt/make-tree)}))]
-
-    (as-> snap-data $
-      (reduce delete-data $ to-delete)
-      (reduce add-frames $ frames-to-add)
-      (reduce add-data $ to-add)
-      (reduce delete-frames $ frames-to-delete))))
-
-;; (defn- log-state
-;;   "Helper function to print a friendly version of the snap tree. Debugging purposes"
-;;   []
-;;   (let [process-frame-data #(d/mapm rt/as-map %)
-;;         process-page-data  #(d/mapm process-frame-data %)]
-;;     (js/console.log "STATE" (clj->js (d/mapm process-page-data @state)))))
-
-(defn- index-page [state page-id objects]
-  (let [snap-data (initialize-snap-data objects)]
-    (assoc state page-id snap-data)))
-
-(defn- update-page [state page-id old-objects new-objects]
-  (let [snap-data (get state page-id)
-        snap-data (update-snap-data snap-data old-objects new-objects)]
-    (assoc state page-id snap-data)))
-
 ;; Public API
 (defmethod impl/handler :snaps/initialize-index
   [{:keys [data] :as message}]
-  ;; Create the index
-  (letfn [(process-page [state page]
-            (let [id      (:id page)
-                  objects (:objects page)]
-              (index-page state id objects)))]
-    (swap! state #(reduce process-page % (vals (:pages-index data))))
-    ;; (log-state)
-    ;; Return nil so the worker will not answer anything back
-    nil))
 
-(defmethod impl/handler :snaps/update-index
-  [{:keys [page-id old-objects new-objects] :as message}]
-  (swap! state update-page page-id old-objects new-objects)
+  (let [pages (vals (:pages-index data))]
+    (reset! state (reduce sd/add-page (sd/make-snap-data) pages)))
 
-  ;; Uncomment this to regenerate the index everytime
-  #_(swap! state index-page page-id new-objects)
-  ;; (log-state)
+  nil)
+
+(defmethod impl/handler :snaps/update-index
+  [{:keys [old-page new-page] :as message}]
+  (swap! state sd/update-page old-page new-page)
   nil)
 
 (defmethod impl/handler :snaps/range-query
-  [{:keys [page-id frame-id coord ranges] :as message}]
-  (letfn [(calculate-range [[from to]]
-            (-> @state
-                (get-in [page-id frame-id coord])
-                (rt/range-query from to)))]
-    (->> ranges
-         (mapcat calculate-range)
-         set ;; unique
-         (into []))))
+  [{:keys [page-id frame-id axis ranges] :as message}]
+  (->> ranges
+       (mapcat #(sd/query @state page-id frame-id axis %))
+       (set) ;; unique
+       (into [])))
```

Comment:
I think this can be implemented using transducers:

```clojure
(into [] (comp #(sd/query @state page-id frame-id axis %) (distinct)) ranges)
```

Reply:
ok